dill/_dill.py

```diff
@@ -1599,46 +1627,40 @@ def save_type(pickler, obj, postproc_list=None):
-            if pickler_is_dill and not pickler._byref and id(obj) not in pickler._postproc:
+    else:
+        obj_name = getattr(obj, '__qualname__', getattr(obj, '__name__', None))
+        _byref = getattr(pickler, '_byref', None)
+        obj_recursive = id(obj) in getattr(pickler, '_postproc', ())
+        if not _byref and not obj_recursive and not _locate_function(obj): # not a function, but the name was held over
+            if issubclass(type(obj), type):
                 # thanks to Tom Stepleton pointing out pickler._session unneeded
                 _t = 'T2'
                 log.info("%s: %s" % (_t, obj))
                 _dict = _dict_from_dictproxy(obj.__dict__)
-        #   except: # punt to StockPickler (pickle by class reference)
             else:
-                warnings.warn('The byref setting is on, but %s cannot be located.' % (obj,), RuntimeWarning)
-                log.info("T5: %s" % obj)
-                name = getattr(obj, '__qualname__', getattr(obj, '__name__', None))
-                StockPickler.save_global(pickler, obj, name=name)
-                log.info("# T5")
-                return
+                _t = 'T3'
+                log.info("%s: %s" % (_t, obj))
+                _dict = obj.__dict__
+           #print (_dict)
+           #print ("%s\n%s" % (type(obj), obj.__name__))
+           #print ("%s\n%s" % (obj.__bases__, obj.__dict__))
+            for name in _dict.get("__slots__", []):
+                del _dict[name]
+            _save_with_postproc(pickler, (_create_type, (
+                type(obj), obj_name, obj.__bases__, _dict
+            )), obj=obj, postproc_list=postproc_list)
+            log.info("# %s" % _t)
         else:
-            _t = 'T3'
-            log.info("%s: %s" % (_t, obj))
-            _dict = obj.__dict__
-       #print (_dict)
-       #print ("%s\n%s" % (type(obj), obj.__name__))
-       #print ("%s\n%s" % (obj.__bases__, obj.__dict__))
-        for name in _dict.get("__slots__", []):
-            del _dict[name]
-        name = getattr(obj, "__qualname__", obj.__name__)
-        _save_with_postproc(pickler, (_create_type, (
-            type(obj), name, obj.__bases__, _dict
-        )), pickler_is_dill, obj=obj, postproc_list=postproc_list)
-        log.info("# %s" % _t)
-    else:
-        log.info("T4: %s" % obj)
-       #print (obj.__dict__)
-       #print ("%s\n%s" % (type(obj), obj.__name__))
-       #print ("%s\n%s" % (obj.__bases__, obj.__dict__))
-        name = getattr(obj, '__qualname__', getattr(obj, '__name__', None))
-        StockPickler.save_global(pickler, obj, name=name)
-        log.info("# T4")
+            log.info("T4: %s" % obj)
+            if _byref:
+                warnings.warn('The byref setting is on, but %s cannot be located.' % (obj,), PicklingWarning)
+            if obj_recursive:
+                warnings.warn('%s.__dict__ contains %s. Cannot pickle recursive classes.' % (obj_name, obj), PicklingWarning)
```

Comment:
Similar comment to above... `warnings.warn("Can't pickle: %s.__dict__ contains %s with recursive self-references." % (obj_name, obj), PicklingWarning)`, or something similar.

Also, for the warning just above: `warnings.warn('Cannot locate reference to %s.' % (obj,), PicklingWarning)`

Reply:
@anivegesana: same comment as above with regard to "perfectly" and "would".